### ISSUE TITLE
How to define the API for parameter initialization, regularization (L2, weight dropout, etc), maybe updater opts per-param

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn_common#59]
It is maybe not such a nice idea if every new `Module` will have this option and then explicitly passes it on to all submodules. Some modules might also not implement this. And maybe there are other options which are not handled yet, e.g. param noise, etc.

How should this be handled? Mostly such arguments which every layer can potentially accept. Mostly arguments about certain behavior on parameters (variables).

In TF, the natural way would be to use `tf.variable_scope` context scope and there have some custom getter or other custom logic.

So maybe also some context manager here?

**Comment by albertz** [comment]
Also `Lstm` with `rec_weight_dropout`. That should be removed, and we should have a generic `weight_dropout`.

**Comment by albertz** [comment]
The same also for parameter initialization options. Like `forward_weights_init` and `bias_init`.

**Comment by albertz** [comment]
`LinearLayer` also has `filter_grad`. This is applied after the matrix multiplication, before the optional activation function.
This should be more generic as well. Basically this could be applied after every layer.
There could be a RETURNN layer for this, like `FilterGradLayer`, which would be somewhat similar to `ScaledGradientLayer`.
Or this could be a generic option for `LayerBase`. And then applied in `LayerBase.post_init`.

Then this becomes also relevant here.

**Comment by albertz** [comment]
`LayerBase` also has `trainable`. Should we also handle that here?

**Comment by albertz** [comment]
`LayerBase` also has `updater_opts`. Should we also handle that here?

**Comment by albertz** [comment]
`LayerBase` also has `param_variational_noise`, which should be handled here.

**Comment by albertz** [comment]
There are some other potential scope-based or global options (all potential `LayerBase` options) like `param_device`.
PyTorch also has this, like `device` or `dtype` (e.g. to use `bfloat16`).

**Comment by albertz** [comment]
Param init, regularization (per param) in most frameworks are options to the corresponding module or layer:

* PyTorch
* Keras
* RETURNN

In lower level TF, param init and regularization is via a context manager (`tf.variable_scope`).

* JAX?
* Sonnet?

**Comment by albertz** [comment]
For param int: No matter if via direct argument, context scope or some other way, there need to be some sensible default, and sometimes also explicit overwrite.

The question is, should there be an easy way to overwrite the defaults globally? This could make sense to easily try out some new kind of param initialization. This could be via a parent context scope, or maybe like the PyTorch `Module.apply` logic.

The problem is, such global overwrite might not make sense for every parameter. Some parameters have other custom default init logic.

How to differentiate those from the standard ones?

One possibility is maybe some clustering of types of init, e.g. for linear matrix, bias, and maybe other types. And then you could globally overwrite the default per type. Although this is tricky and maybe hard to get right and will likely never fit all cases.

Also, if there are multiple (maybe nested) settings of defaults, which one will take precedence? The most recent one or the first (most outer) one? There are reasons for priority on the most outer one to be able to globally overwrite this, otherwise such outer settings would never have an effect. But there are maybe also cases where some inner setting should take precedence. Do we need a priority system for this? But that would make it way too complicated.

**Comment by albertz** [comment]
[PyTorch also has optimizer-per-parameter settings](https://pytorch.org/docs/stable/optim.html#per-parameter-options).

**Comment by albertz** [comment]
PyTorch `Module.apply` has even an example in its documentation on using it to overwrite param init.

`Module.apply` together with a forward pre-hook is also used to implement weight normalization. See [here](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/weight_norm.py). The forward pre-hook calculates the original weights. In TF/RETURNN, such a pre-hook would not really be needed.

So maybe such `apply` API is a nice and simple solution for us as well.

Although recently there has been some discussion on PyTorch that this is actually not so nice for reparameterization including weight norm, and a new API for reparameterization was proposed, discussed and implemented: https://github.com/pytorch/pytorch/issues/7313, https://github.com/pytorch/pytorch/issues/28937, https://github.com/pytorch/pytorch/pull/33344, [parametrizations tutorial](https://pytorch.org/tutorials/intermediate/parametrizations.html).

I'm not sure though that the new PyTorch parametrization API makes it really cleaner and simpler. While it looks nice from a high-level viewpoint, it hides many things.

**Comment by albertz** [comment]
Note, our parameters are now always `nn.Parameter` instances, which derive from `Layer`, and thus `LayerRef`.

Maybe they should have some `initial` attribute.
(The implementation behind that would require some extra effort. We might want to map this to the `init` layer option of the underlying `VariableLayer`. However, we can not pass other layer refs there! So this needs some extension on RETURNN side. I'm not exactly sure how difficult this would be.)

**Comment by albertz** [comment]
When it is easy to overwrite the param init directly via `module.param.initial = ...`, do we even need the param init as an argument to modules? Modules could just define some sensible default init and this could be overwritten later explicitly.

**Comment by albertz** [comment]
L2 on weights would probably also be done via `apply`. And this should be straight-forward using `mark_as_loss` (#56).

**Comment by albertz** [comment]
Weight dropout could just be done like weight norm explained above.

**Comment by albertz** [comment]
I wonder whether we need the param init as an argument to the modules at all. With such argument, there are two ways to overwrite the default:
```
linear = Linear(..., weight_init=...)
```
Vs:
```
linear = Linear(...)
linear.weight.init = ...
```

But the second variant would be possible in any case. So, do we really need the first way then?

**Comment by albertz** [comment]
Things like param init, weight dropout, weight norm, L2 on weights, etc can all easily be handled in such post-processing way, and recursively via such `apply` API. So this doesn't need to be an argument of the module.

I also don't like it when there are too many arguments. So I would prefer to keep the argument list short.

Context managers are an alternative to arguments. Context managers might be added for param init or other regularization at some point (but not really needed; and I would only do that if the behavior would be simple and straight-forward). Context managers are probably the way to go for `dtype` and `device`.

One remaining question are options for model-based regularization such as dropout (on activations). Dropout is by far the most common thing but there might be other variants, and a complex model might also have various options where dropout could be added. If this is configured via model argument, this could easily become bloated when the model becomes complex or when there are other types of regularization in the future. So I also don't like this. Context managers might be somehow tricky here as well, to get this right with straight-forward behavior. However, we can maybe again do this via post-processing. Maybe modules could provide functions like `set_dropout`. Or maybe `nn.scoped` could return an object which allows for functions like `register_post_process` where dropout could be added.

**Comment by albertz** [comment]
Another model-based regularization is stochastic depth. This can be implemented efficiently using `Cond` (#24) for `train_flag() && random_uniform() > 0.5` as condition or so.

But the question here for this issue is again, how would the option look like? An argument to the module? Or some module method like `set_stochastic_depth`? Or via `register_forward_hook` on individual sub modules?

**Comment by albertz** [comment]
Note that param init also requires some further thought on the technical side: #92

**Comment by albertz** [comment]
A lot of relevant similar discussion on the API how to define the param init is here: https://github.com/pytorch/pytorch/issues/18182

**Comment by albertz** [comment]
Related is also what param init defaults we should use. See #94.

**Comment by albertz** [comment]
Related is [TF arg_scope](https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/contrib/framework/python/ops/arg_scope.py) which provides a context scope to overwrite the defaults for kwargs like `initializer` or `regularizer`, [example from OpenAI weightnorm](https://github.com/openai/weightnorm/blob/3503e68b84d27ac69f33e26517f2a5036153999c/tensorflow/nn.py). But I have not really seen much adoption of this and I think it was removed in TF 2.

**Comment by albertz** [comment]
Related is also #96 on explicit training loop and stages. When we implement that, the param init would be outside of the training loop, and probably it becomes all more clear.

**Comment by albertz** [comment]
I think we should leave this open until this is really implemented, or until we maybe have a separate issue specifically on the technical aspect.

**Comment by albertz** [comment]
`Module.apply` is implemented now. This allows for custom param init and many other things mentioned here.

L2 on params is still another aspect. We could just do `sum(nn.reduce_sum(param ** 2) for param in net.parameters()).mask_as_loss()`. However, we know from literature (e.g. AdamW) that it is better to have this decoupled from the loss optimization, and not use momentum for this part. Because we don't really want L2, we want weight decay. Weight decay is equivalent to L2 only with SGD. The cleanest/simplest approach would probably be sth like explicit custom updates, like #90.

**Comment by albertz** [comment]
For L2 on params, we could also use the RETURNN mechanism directly as an option to the `VariableLayer`.

**Comment by albertz** [comment]
The `Parameter.weight_decay` property is implemented now.

I think we can close this here for now.
